DIV-5762 Removed unwanted integration tests moved some to nightly

hosainnet opened the pull request:
# Description

To reduce pipeline flakyness and speed up build
Effectively testing the main integration/happy path every PR, and other scenarios nightly

Fixes #DIV-5762



## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)


# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/callback/CoRespReceivedCallbackTest.java:
```diff
@@ -30,20 +26,4 @@ public void givenUndefendedCase_whenSubmitAOS_thenReturnAOSData() {
-
-        assertEquals(1,error.size());
-        assertTrue(error.get(0).contains("email_address is a required property"));
-    }
```
Moved to CoRespondentSubmittedITest.java  below

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/callback/DNReceivedTest.java:
```diff
@@ -36,21 +35,4 @@ public void givenCase_whenDNSubmitted_thenReturnDNDataPlusAnswerDoc() {
-            is(DOCUMENT_TYPE_DN_ANSWERS)));
-        assertThat(jsonResponse, hasJsonPath("$.data.D8caseReference",
-            is("LV17D80100")));
-    }
```
`givenNotificationError_whenExecuteEmailNotificationTask_thenReturnData`

charlesluokb replied:
what does the above comment mean? ☝️

hosainnet replied:
That's the alternative unit test that is covering this @charlesluokb

hosainnet replied:
@charlesluokb sorry I forgot to comment here, but it replaced by a test with the same name below
https://github.com/hmcts/div-case-orchestration-service/pull/507/files#diff-32886a355a4f73efdc64823ad82bdbceR115

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/callback/PetitionIssueTest.java:
```diff
@@ -71,26 +66,6 @@
-        assertEquals(HttpStatus.OK.value(), cosResponse.getStatusCode());
-        assertGeneratedDocumentsExists(cosResponse, false, false);
-        assertEquals(EXPECTED_ISSUE_DATE, cosResponse.path(ISSUE_DATE));
-    }
```
We never pass "false" as a flag to `generateAos` GET param

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/callback/PetitionIssueTest.java:
```diff
@@ -71,26 +66,6 @@
-        assertGeneratedDocumentsExists(cosResponse, false, false);
-        assertEquals(EXPECTED_ISSUE_DATE, cosResponse.path(ISSUE_DATE));
-    }
-
```
This is testing the non-service centre flow which doesn't exist anymore

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/dataextraction/FamilyManDataExtractionIT.java:
```diff
@@ -26,6 +27,7 @@
     @Value("${case.orchestration.jobScheduler.extract-data-to-family-man.context-path}")
     private String testDataExtractionEndPoint;
 
+    @Ignore //flaky test
```
This fails way too often ATM, will need further investigation

charlesluokb replied:
providing some message as `@ignore("some info goes here")` would be handy

hosainnet replied:
Sorry not sure what else I can provide? in my experince it failed everytime locally so it's easy to reproduce
Hopefully this is very short-term and we can fix it soon

charlesluokb replied:
do you know when you will re-visit this test? leave some comments here suggesting what to do next so that it will be easier for  other developer to pick up when visiting this area.

jvictorcf replied:
This has to run with MailHog working on the right port if it's locally. I've stated that in the comments but if it's not clear, I can make it better.
Please un-comment this as soon as possible, @hosainnet .

jvictorcf replied:
The idea here is that you can run MailHog, un-ignore these tests, run the tests and check your assertions manually.
It shouldn't be failing if it's ignored.
Maybe I don't fully understand the issue.

hosainnet replied:
@jvictorcf sorry I think I missed the comment because it wasn't immediately above the test..
2 quick points
1) We should not have to manual steps to run an automated test (apart from the intial config/secrets change of course)
2) This test has been very flaky on Jenkins, see:
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/331/console
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/330/console
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/328/
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/327/console
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/325/console
https://build.platform.hmcts.net/job/HMCTS_DIV/job/div-case-orchestration-service/job/master/324/console
That's 6 failures in the last 10 master builds... (not counting PRs)

jvictorcf replied:
No worries.
On your point:
1 - you're right. this is a tech debt and it should be addressed.
2 - this was not flaky in the pipeline. it was actually failing and the fix was this PR: https://github.com/hmcts/div-case-orchestration-service/pull/501

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/LinkRespondentTest.java:
```diff
@@ -98,81 +98,6 @@ public void givenAosAwaitingState_whenLinkRespondent_thenCaseShouldBeLinked() {
     }
 
-    @Test
-    public void givenAosOverdueState_whenLinkRespondent_thenCaseShouldBeLinked() {
```
Covered by `givenAosAwaitingState_whenLinkRespondent_thenCaseShouldBeLinked` above

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/LinkRespondentTest.java:
```diff
@@ -98,81 +98,6 @@ public void givenAosAwaitingState_whenLinkRespondent_thenCaseShouldBeLinked() {
-    }
-
-    @Test
-    public void givenValidCaseDetails_whenLinkCoRespondent_thenCaseShouldBeLinked() {
```
Covered by givenLinkedCase_whenLinkCoRespondent_thenCaseShouldBeLinked below

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/RetrieveAosTest.java:
```diff
@@ -25,13 +25,6 @@
-
-        assertEquals(HttpStatus.NOT_FOUND.value(), cosResponse.getStatusCode());
-    }
-
```
Covered by `givenNoCaseExists_whenRetrieveAosCase_thenReturnEmptyResponse`

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/RetrieveCaseTest.java:
```diff
@@ -25,13 +25,6 @@
-
-        assertEquals(HttpStatus.NOT_FOUND.value(), cosResponse.getStatusCode());
-    }
-
```
Covered by givenNoCaseExists_whenGetCase_thenReturnEmptyResponse

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitCaseToCCDIntegrationTest.java:
```diff
@@ -45,46 +44,6 @@ public void givenDivorceSession_WithNoCourt_whenSubmitIsCalled_CaseIsCreated() t
-
-        ResponseBody retrieveCaseResponseBody = retrieveCase(userDetails.getAuthToken()).body();
-        assertThat(retrieveCaseResponseBody.path(RETRIEVED_DATA_COURT_ID_KEY), is(allocatedCourt));
-    }
```
The above test made no sense - no idea where `unknown-court` will comes from so that test will always pass

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitCaseToCCDIntegrationTest.java:
```diff
@@ -45,46 +44,6 @@ public void givenDivorceSession_WithNoCourt_whenSubmitIsCalled_CaseIsCreated() t
-        allocatedCourt = caseCreationResponseBody.path(ALLOCATED_COURT_ID_KEY);
-        assertThat(allocatedCourt, is(notNullValue()));
-    }
-
```
Covered by `givenCaseAlreadyExists_whenCaseDataIsSubmitted_thenReturnSuccess`

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitCoRespondentAosCaseTest.java:
```diff
@@ -58,22 +56,6 @@
-
-        assertThat(coRespondentSubmissionResponse.getStatusCode(), is(BAD_REQUEST.value()));
-    }
-
```
Covered by `SubmitCoRespondentAosCaseUTest. willThrowExceptionIfCaseInWrongState`

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitDaCaseTest.java:
```diff
@@ -29,17 +25,7 @@
-        } catch (FeignException exception) {
-            assertEquals(BAD_REQUEST.value(), exception.status());
-        }
-    }
```
Covered by `givenNoAuthToken_whenSubmitDa_thenReturnBadRequest`

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitDnCaseTest.java:
```diff
@@ -7,19 +7,11 @@
-
-        assertEquals(BAD_REQUEST.value(), cosResponse.getStatusCode());
-    }
-
```
`givenNoAuthToken_whenSubmitDn_thenReturnBadRequest`

hosainnet commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitRespondentAosCaseTest.java:
```diff
@@ -66,109 +51,6 @@ public void givenConsentAndDefend_whenSubmitAos_thenProceedAsExpected() throws E
     }
 
-    @Test
-    public void givenNoConsentAndDefend_whenSubmitAos_thenProceedAsExpected() throws Exception {
```
The above test is enough to test the integration
The other cases are basically testing the different possible events which are covered by `SubmitRespondentAosCaseUTest`

alexwakeman commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitCaseToCCDIntegrationTest.java:
```diff
@@ -46,24 +47,7 @@ public void givenDivorceSession_WithNoCourt_whenSubmitIsCalled_CaseIsCreated() t
-
-    @Test
+    @Category(NightlyTest.class)
     public void givenAnExistingCase_whenSubmitIsCalled_aNewCaseIsNotCreated() throws Exception {
```
I think we'd only need this one in the CMS right? I guess as Nightly it's fine

hosainnet replied:
I think the logic which checks for an existing case in the first place is in COS

alexwakeman replied:
It's all good, but the logic is in CMS: https://github.com/hmcts/div-case-maintenance-service/blob/master/src/main/java/uk/gov/hmcts/reform/divorce/casemaintenanceservice/service/impl/CcdRetrievalServiceImpl.java#L103

alexwakeman replied:
Sorry, you might be right there - somehow I think it *should* be in the CMS.

alexwakeman commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/SubmitCoRespondentAosCaseTest.java:
```diff
@@ -93,6 +96,7 @@ public void canSubmitAndRetrieveCoRespondentAnswers() throws Exception {
     }
 
     @Test
+    @Category(NightlyTest.class)
```
👍 nice

charlesluokb commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/callback/GetPetitionIssueFeesTest.java:
```diff
@@ -71,32 +68,4 @@ public void givenCallbackRequest_whenGetPetitionIssueFees_thenReturnUpdatedData(
     }
-
-    @Test
-    public void givenUnauthorizedRequest_whenGetPetitionIssueFees_thenReturnErrorData() {
```
why this test is removed?

alexwakeman replied:
We can test this scenario using functional tests, given we don't test for error response status code. We're leaning toward happy path integration tests.

hosainnet replied:
@charlesluokb sorry I forgot to comment here, but it replaced by a test with the same name below
https://github.com/hmcts/div-case-orchestration-service/pull/507/files#diff-32886a355a4f73efdc64823ad82bdbceR115

charlesluokb commented on src/integrationTest/java/uk/gov/hmcts/reform/divorce/maintenance/LinkRespondentTest.java:
```diff
@@ -137,6 +141,7 @@ public void givenAosOverdueState_whenLinkRespondent_thenCaseShouldBeLinked() {
     }
 
     @Test
+    @Category(NightlyTest.class)
```
Nice to use @Category 👍

yarpo commented:
@jvictorcf - the think is we are unable to build anything on jenkins. We need to make a proper review of the amount of integration tests and the ways of speeding them up.

Having a bunch of tests that never pass (build has been taking too long) is not helpful.

I'm happy to discus it.

hosainnet commented:
> @hosainnet, why should these tests only be run nightly? They're part of the core business, aren't they?

We're covering the happy path at least for every 3rd party integration. The flaykness of the other envs are making the coverage of all possible paths by integration impossible
Until we see a significant improvement in those services, we're wasting 100s of man hours rebuilding PRs and master

jvictorcf commented:
I agree this is a problem, but aren't we turning a blind eye to our own potential failures in this way?
I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.

hosainnet commented:
>  I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.

This won't stop your running all integration tests locally before you push (we're already doing this, correct? ;))

jvictorcf commented:
> > I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.
> 
> This won't stop your running all integration tests locally before you push (we're already doing this, correct? ;))

I don't we should count on this manual step. This would be trusting people rather than process.
Of course I do understand and feel the pressure to deliver and the motivation behind this. I'm just not convinced it's the right approach, but again, I might be missing an angle here.

hosainnet commented:
> > > I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.
> > 
> > 
> > This won't stop your running all integration tests locally before you push (we're already doing this, correct? ;))
> 
> I don't we should count on this manual step. This would be trusting people rather than process.
> Of course I do understand and feel the pressure to deliver and the motivation behind this. I'm just not convinced it's the right approach, but again, I might be missing an angle here.

@jvictorcf test doubles are already exist in the form of the IT tests we have in  https://github.com/hmcts/div-case-orchestration-service/tree/master/src/test/java/uk/gov/hmcts/reform/divorce/orchestration/functionaltest (assuming you mean mocking the services)

Togther with unit tests, those provide 96% coverage (excluding the integration tests) https://sonarcloud.io/code?id=div-case-orchestration-service
Integration tests don't count towards coverage, they're just ensuring we can connect to the services (at least..)
In the past 3 months, we:
1) Added retry for errors while trying to create users: https://github.com/hmcts/div-case-orchestration-service/blob/5f20e3947b67d17f4626dbfb46eb9064189ff972/src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java#L142
2) Added a retry rule to repeat failing test 3 times https://github.com/hmcts/div-case-orchestration-service/blob/master/src/integrationTest/java/uk/gov/hmcts/reform/divorce/RetryRule.java
3) Communicated with teams _many_ times but the underlying issues were never fixed
```
(22nd Aug) https://tools.hmcts.net/jira/browse/SIDM-3061 
https://hmcts-reform.slack.com/archives/CAT2XTU9J/p1569495974198600
```

If it was up to me, I'd run all tests as integration - because ultimately that's the only way we can ensure all services are configured properly. But currently some services we rely on don't really treat this instability problem as a priority - so given all this data above I'm really interested to know other approaches to this

jvictorcf commented:
> > > > I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.
> > > 
> > > 
> > > This won't stop your running all integration tests locally before you push (we're already doing this, correct? ;))
> > 
> > 
> > I don't we should count on this manual step. This would be trusting people rather than process.
> > Of course I do understand and feel the pressure to deliver and the motivation behind this. I'm just not convinced it's the right approach, but again, I might be missing an angle here.
> 
> @jvictorcf test doubles are already exist in the form of the IT tests we have in https://github.com/hmcts/div-case-orchestration-service/tree/master/src/test/java/uk/gov/hmcts/reform/divorce/orchestration/functionaltest (assuming you mean mocking the services)
> 
> Togther with unit tests, those provide 96% coverage (excluding the integration tests) https://sonarcloud.io/code?id=div-case-orchestration-service
> Integration tests don't count towards coverage, they're just ensuring we can connect to the services (at least..)
> In the past 3 months, we:
> 
> 1. Added retry for errors while trying to create users: https://github.com/hmcts/div-case-orchestration-service/blob/5f20e3947b67d17f4626dbfb46eb9064189ff972/src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java#L142
> 2. Added a retry rule to repeat failing test 3 times https://github.com/hmcts/div-case-orchestration-service/blob/master/src/integrationTest/java/uk/gov/hmcts/reform/divorce/RetryRule.java
> 3. Communicated with teams _many_ times but the underlying issues were never fixed
> 
> ```
> (22nd Aug) https://tools.hmcts.net/jira/browse/SIDM-3061 
> https://hmcts-reform.slack.com/archives/CAT2XTU9J/p1569495974198600
> ```
> 
> If it was up to me, I'd run all tests as integration - because ultimately that's the only way we can ensure all services are configured properly. But currently some services we rely on don't really treat this instability problem as a priority - so given all this data above I'm really interested to know other approaches to this

Hello @hosainnet. Could one option be to have external services as chart dependencies?
I'm all in favour of having a session or two where we can collectively come up with ideas to improve our current situation.

hosainnet commented:
> > > > > I think I'd rather have test doubles of external dependencies and still run the tests that cover our functionalities in every commit.
> > > > 
> > > > 
> > > > This won't stop your running all integration tests locally before you push (we're already doing this, correct? ;))
> > > 
> > > 
> > > I don't we should count on this manual step. This would be trusting people rather than process.
> > > Of course I do understand and feel the pressure to deliver and the motivation behind this. I'm just not convinced it's the right approach, but again, I might be missing an angle here.
> > 
> > 
> > @jvictorcf test doubles are already exist in the form of the IT tests we have in https://github.com/hmcts/div-case-orchestration-service/tree/master/src/test/java/uk/gov/hmcts/reform/divorce/orchestration/functionaltest (assuming you mean mocking the services)
> > Togther with unit tests, those provide 96% coverage (excluding the integration tests) https://sonarcloud.io/code?id=div-case-orchestration-service
> > Integration tests don't count towards coverage, they're just ensuring we can connect to the services (at least..)
> > In the past 3 months, we:
> > 
> > 1. Added retry for errors while trying to create users: https://github.com/hmcts/div-case-orchestration-service/blob/5f20e3947b67d17f4626dbfb46eb9064189ff972/src/integrationTest/java/uk/gov/hmcts/reform/divorce/context/IntegrationTest.java#L142
> > 2. Added a retry rule to repeat failing test 3 times https://github.com/hmcts/div-case-orchestration-service/blob/master/src/integrationTest/java/uk/gov/hmcts/reform/divorce/RetryRule.java
> > 3. Communicated with teams _many_ times but the underlying issues were never fixed
> > 
> > ```
> > (22nd Aug) https://tools.hmcts.net/jira/browse/SIDM-3061 
> > https://hmcts-reform.slack.com/archives/CAT2XTU9J/p1569495974198600
> > ```
> > 
> > 
> > If it was up to me, I'd run all tests as integration - because ultimately that's the only way we can ensure all services are configured properly. But currently some services we rely on don't really treat this instability problem as a priority - so given all this data above I'm really interested to know other approaches to this
> 
> Hello @hosainnet. Could one option be to have external services as chart dependencies?
> I'm all in favour of having a session or two where we can collectively come up with ideas to improve our current situation.

That's a good suggestion. I just checked though, and for example, https://github.com/hmcts/chart-idam is not in a usable state - (we also need to create product charts out of all our apps)
